Initial implementation of configuration actions

cdjackson opened the pull request:
As per the discussion [here](https://www.eclipse.org/forums/index.php/t/1066252/)

This is an initial suggestion for configuration actions. The actions are intended as immediate 'commands' for Thing configuration purposes - for example, to reset a device, exclude it from the network, perform some other function needed to initialise/configure/administer a device.

As per the discussion on the Eclipse forum, my current use case, and what I implemented for HABmin in OH1 was just a simple action::value command - basically implemented as buttons. I've implemented this similarly to parameters - you might consider that overkill, and I might agree, but I can also see that someone might come along with a use case for something a little more detailed than a simple button...

In the XML, in the parameter-descriptions, I've added ```<action>```. This builds the action list alongside the parameters and can be retrieved as part of the config registry and hence into the REST interface for the thing type. It adds a ```actions{}``` element to the REST json which lists all the actions.

In the other direction, I've added a thing/action interface to REST. There's then a ThingConfigAction which is expected to be implemented by the ThingHandler. This has a single method ```handleConfigurationAction(Thing thing, String action, String value)```.

I've not tested the sending of actions yet - only the XML and definitions into the REST interface. Before I go too far I wanted to see if you're happy with this approach...

The questions I can see are -:
* If the action definition is a bit too much (filters could probably be removed, maybe some other stuff)
* If the types are right for an action - maybe it should be more description like BUTTON or something. However in the API I would potentially ignore this anyway as this can probably be established in other ways.

Comments welcome :)

Signed-off-by: Chris Jackson <chris@cd-jackson.com>

AlexTugarev commented on addons/binding/org.eclipse.smarthome.binding.yahooweather/ESH-INF/thing/thing-types.xml:
```diff
@@ -27,6 +27,18 @@
                 <default>60</default>
             </parameter>
+
+            <action name="location" type="text">
```
I'd suggest to use verbs for naming of action.

AlexTugarev replied:
Now, I got it!
This says: action should be named like the the parameter/-group it's going to provide edit support for.

Have a look at the redundancy this solution would ends in: 
```
        <config-description>
            <parameter name="location" type="text" required="true">
                <label>Location</label>
                <description>Location for the weather information.
                    Syntax is WOEID, see https://en.wikipedia.org/wiki/WOEID.
                </description>
            </parameter>
            <parameter name="refresh" type="integer">
                <label>Refresh interval</label>
                <description>Specifies the refresh interval in seconds.</description>
                <default>60</default>
            </parameter>

            <action name="location" type="text">
                <label>Location</label>
                <description>Location for the weather information.
                    Syntax is WOEID, see https://en.wikipedia.org/wiki/WOEID.
                </description>
            </action>
            <action name="refresh" type="integer">
                <label>Refresh interval</label>
                <description>Specifies the refresh interval in seconds.</description>
                <default>60</default>
            </action>
        </config-description>
    </thing-type>
```

All the labels, descriptions, and other attributes are duplicated. 
Can't it be expressed like this?
```
<config-description>
    <parameter name="location" type="text" required="true">
        <label>Location</label>
        <description>...</description>
    </parameter>

    <action name="updateLocation">
        <label>Update Location</label>
        <parameters>
           <parameter>location</parameter>
        </parameters >
    </action>
</config-description>
```

cdjackson replied:
> action should be named like the the parameter/-group it's going to provide edit support for

NO - actions and parameters are not intended to be linked in the way you are thinking. You would have a bunch of parameters, and some actions.

You are looking at the test configuration where I just copied the same XML from parameters into action, so yes, there is a lot of redundancy. That's clearly not the intention in real life - this is just the test file. Sorry if this confused matters, but I didn't make the test case real-life, it was just there to test the functionality (sorry).

As I said earlier, actions are there to provide an immediate action - you wouldn't have an action and a parameter that are the same - eg you wouldn't have a reset parameter, and a reset action as that doesn't make sense?

I would expect that a device has a small number of actions where the user needs to trigger something, and a number of parameters where values are set. This is how it is in the zwave binding currently.

dnobel commented:
I´m not convinced about this concept at all. We already have the problem that binding developer can not distinguish between channels, configuration and properties. By introducing such a generic concept as the proposed action, I think it will be even more unclear. If the concept of actions is related to the configuration, maybe it should be part of the config description part. But we should definitely name it "config-action", otherwise it is very confusing and can be misinterpreted as some kinds of commands.

However, if we decided to include it, we should reference a config description in the schema, instead of reintroducing all the elements from the config description.

AlexTugarev commented:
@dnobel I don't get the point about things will be more unclear, if config parameters will be enriched with editing semantics. #editing-semantics

AlexTugarev commented:
@dnobel @cdjackson I'd prefer to move the description of config-action to a separate description resource. Such config-actions would be reusable, if you think on chaining those.

AlexTugarev commented on bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/thing/ThingResource.java:
```diff
@@ -199,6 +200,28 @@ public Response update(@PathParam("thingUID") String thingUID, ThingBean thingBe
+            return Response.status(Status.NOT_FOUND).build();
+        }
+
+        if(thing instanceof ThingConfigAction) {
```
'Would prefer a non-invasive design w/o touching core.thing at all. 'Thinking of a service for config-action to ask for, and trigger, iff applicable.

AlexTugarev commented on bundles/config/org.eclipse.smarthome.config.xml/config-description-1.0.0.xsd:
```diff
@@ -78,6 +81,33 @@
+		<xs:attribute name="step" type="xs:decimal" />
+		<xs:attribute name="pattern" type="xs:string" />
+		<xs:attribute name="multiple" type="xs:boolean" />
+	</xs:complexType>
```
How are these actions meant to be mapped to the config-parameters?

dnobel commented:
@AlexTugarev Sorry, you are right. I have read the XML file wrongly. The actions are already part of the config description. So it should be ok. But we should somehow combine the parameter and action elements to a single concept.

AlexTugarev commented on bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/thing/ThingResource.java:
```diff
@@ -199,6 +200,28 @@ public Response update(@PathParam("thingUID") String thingUID, ThingBean thingBe
+        }
+
+        if(thing instanceof ThingConfigAction) {
+            ((ThingConfigAction)thing).handleConfigurationAction(thing, actionId, actionValue);
```
In case of long-running actions, this sync call would lead to comm issues.

AlexTugarev commented:
@dnobel Actions being part of the config-descriptions is not a nice solution FMPOV. If thing of embedding the action in to a ThingType, you'll miss the change to reuse the same description for further ThingTypes. Instead of extending "config.xml", wich is also used by other description types, I'd prefer to introduce a new description type for these config-action. XML instances of config-action description would then point to ThingTypes by id, which means, that these actions can provide #editing support of Things on instance-level only.

cdjackson commented:
@AlexTugarev my view was that the actions were part of the configuration interfaces - we define parameters in the config description, and it seemed sensible (to me at least :) ) to combine them at the same point. They are therefore reusable in the same was as config descriptions - I didn't consider that someone might want to use them separately for some reason as to me they are linked to the thing.

What's the use case for using the config-actions separately to the configuration descriptions?

AlexTugarev commented:
@cdjackson Fact is config-descriptions can be defined standalone and referenced from thing-type.xml *and* binding.xml.

kaikreuzer commented:
@AlexTugarev Yes, and config-descriptions can also be used for new rule modules (https://github.com/eclipse/smarthome/pull/317) and all kinds of other services.
But why is this a reason to have config-actions outside of the config-descriptions?

cdjackson commented:
It's the same here though - it's part of the config-descriptions so can be used anywhere that the config descriptions are used.

AlexTugarev commented:
@kaikreuzer @cdjackson Attaching config-action to config-descriptions (instead of ThingTypes) implies that not only Things are configurable. Also reusing of config-actions across several ThingTypes with a common set of parameters is difficult and likely limited to use case with a common config-description.

cdjackson commented:
> Attaching config-action to config-descriptions (instead of ThingTypes) implies that not only Things are configurable.

But doesn't the fact that it's called a 'config-description' mean that it's configurable anyway? Anything that supports the config-description is presumably configurable?



> Also reusing of config-actions across several ThingTypes with a common set of parameters is difficult and likely limited to use case with a common config-description.

In my mind, they are linked - a user might want to reset a thing with an action, and then set a parameter. Just because things and actions are both in the config-description doesn't mean that a thing must support both - the same with groups which are also part of the config-description - it's not mandatory to support them if it doesn't make sense.

AlexTugarev commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/ThingConfigAction.java:
```diff
@@ -0,0 +1,32 @@
+ *
+ * @author Chris Jackson - Initial contribution and API
+ */
+public interface ThingConfigAction {
```
Where is a instance of type ThingConfigAction created?

cdjackson replied:
Currently, on the ThingHandler

AlexTugarev replied:
How? 

I mean, a Thing is associated with a ThingHandler, and in ThingResource you do a (thing instanceof ThingConfigAction). How does this fit together?

cdjackson replied:
As I said in the beginning, I've not tested this part, but the idea what that the ThingHandler simply implements the ThingConfigAction interface.

AlexTugarev replied:
okay, then you probably meant (thing.getHandler instanceof ThingConfigAction), right?

cdjackson replied:
Probably:) I'd need to check (and test!)... The point of this PR was to get some feedback as the discussion in the forum had stalled, so I appreciate the feedback...

I'm in the position where I will need to get this functionality implemented quite soon so if actions are not a concept that people like I'll need to look for alternatives, but I'm not sure how to get similar functionality with the current system...

dnobel commented:
Hi Chris,

sorry for the late response. I just want to give some background, what is going on at QIVICON. The topic config actions is a litte bit more complicated, because we are working on a similar concept at QIVICON at the moment. @AlexTugarev and @rodneyrehm are mainly working on this topic.

We are designing a config flow, that can can be defined by a binding. The concept of config actions is a small part of it. But we need to see how to align the concepts. If possible, we should first complete the topic of config actions, but keeping the more bigger topic flows in mind, so that it can be put on top later.

However, regarding your current proposal for config actions in your PR, I think we should discuss some additional questions:

1) In which case should the binding developer use a config action, and in which case a property?
2) How are config actions and properties related? For example we have the problem with the hue bridge, that the user must press the button after configuring a property. Can this be solved with the concept of config actions?
3) Is it really related to configuration at all? Maybe it is just a functionality of the device, you want to call.
4) How should the UI render config actions in contrast to properties? Do you have an example in HABmin?

We know that this topics blocks the implementation of your z-wave binding and we don´t want to hold you back from contributing this stuff. But it may need some more discussions and clarifications, before we can accept a PR here.

Regards

Dennis

cdjackson commented:
Hi Dennis,
My thought is that config actions (and maybe this name is confusing, so I don't care about the name :) ) are parameters that are updated immediately. 

My immediate requirement, and what I've used this for in OH1/HABmin, is to be able to send a device config command such as a reset, or trigger some action like a network reconfiguration. However, I can also see the use case where numeric, or text values might also need to be sent. Currently with parameters you can't set a value to the same value, and have the binding do something (so if I want to resend the value of 12 to a device, I have to change to 11 first, and then to 12). Again, an action would provide this facility.

Another way to do this (At least in the definition) might be to add an attribute to a config parameter, but we would still need some way to trigger this, in real-time, back to the binding. Currently this isn't possible as the config parameters are only values, so we can't add any attributes and you weren't keen to change this. We would still then need a service, or interface, or ?? on the implementation side...

In my mind, this is linked to configuration - it all allows me to configure a device/thing, so that's why I've put the definition of the action into the config-description. I think they should be linked, but as above there are a few ways to implement this on the java side at least...

I'm happy to continue discussing this, but the discussion stalled on the Eclipse list. Kai was happy that such a concept was needed so I implemented this PR to try and get things moving - I'm happy to implement something else, but I really need to be able to do something soon as I need this in the Zigbee binding and there is some time pressure.  If you already have other concepts, maybe we can discuss, either here or in private if you prefer as I don't want to waste anyones time if there is already a concept planned...

Cheers
Chris

dnobel commented:
Hi Chris,

> be able to send a device config command such as a reset, or trigger some action

I agree that this not possible with config parameters. This is a a real action that can be triggered at the device. I´m just not convinced, that this should be part of the configuration description.

> Currently with parameters you can't set a value to the same value, and have the binding do something (so if I want to resend the value of 12 to a device, I have to change to 11 first, and then to 12)

Why is this not possible with the current configuration? Even if you send the same value, the thingUpdated method on the ThingHandler should be called. But we discussed that maybe not all config options of the device should be persisted in the thing instance.

> Another way to do this (At least in the definition) might be to add an attribute to a config parameter, but we would still need some way to trigger this, in real-time, back to the binding. Currently this isn't possible as the config parameters are only values, so we can't add any attributes and you weren't keen to change this. We would still then need a service, or interface, or ?? on the implementation side...

Still don´t understand why you need an additional attribute. What should it hold? The service is the ThingHandler. If you change a config parameter, the ThingHandler method "thingUpdated" is called. The only disadvantage is, that you can not figure out, which property has been changed. Do you mean this problem?

> In my mind, this is linked to configuration - it all allows me to configure a device/thing

Maybe it is only related to devices, but not to other service configurations or rule modules (where config descriptions are also used). If we see the action as some state independent method like a reset, it could make sense to pull it up the thing type.

> Kai was happy that such a concept was needed so I implemented this PR to try and get things moving

I also agree, that the current config concept is not sufficient for all use cases. So definitely thanks for the contribution. I think having a first API draft gives as the chance to discuss this better - as we do right now.

>  If you already have other concepts, maybe we can discuss, either here or in private if you prefer as I don't want to waste anyones time if there is already a concept planned

The concept of config flows has another focus  and the implementation on that has not been started. It is also in a conceptional discussion phase. It is just related to config actions. So our current challenge is, to see wether is it possible to bring the concepts together, or at least make it possible to build up the config flow concept on the config action concept. But maybe it something completely different.

Regards

Dennis

cdjackson commented:
Hi Dennis,
Thanks for the response. 

> >Currently with parameters you can't set a value to the same value, and have the binding do something (so if I want to resend the value of 12 to a device, I have to change to 11 first, and then to 12)

> Why is this not possible with the current configuration? 

How do you know it has changed? If it was 12 at one instance in time, and 12 the next time the handler is called, how do I know I have to send this update to the device? There's nothing that tells me that there is an update. You could assume that all parameters have to be updated every time, but this might take a long time on battery devices that only wake every hour for a short time. Additionally, I'm not sure that it is a good idea to reconfigure a device in all instances since it might have secondary effects like resetting a counter, or some internal state.

> Still don´t understand why you need an additional attribute. What should it hold?

It could hold that this is an action parameter. Then, instead of sending the update in the config description, it could be used to trigger an update through the config-action interface.

> The only disadvantage is, that you can not figure out, which property has been changed. Do you mean this problem?

Yes - this is what I mean. As above, for some devices, just resending all configuration when you don't want to will reset some internal state, and may take a long time (and generally, is probably a bad idea).


> Maybe it is only related to devices, but not to other service configurations or rule modules (where config descriptions are also used). If we see the action as some state independent method like a reset, it could make sense to pull it up the thing type.

Yes, this is probably correct, but I don't know that we should define an action that is just as simple as a reset. While this would likely meet my current needs, I think it would be better to define a more capable interface or we'll likely be having this discussion again in the future :)

>  I think having a first API draft gives as the chance to discuss this better - as we do right now.

I'm happy to discuss further, and started a discussion on the forum a month ago, but this stalled. I defined what I thought the definition should look like - I didn't define the API for sending action to the thing, but I don't think this is difficult if we can decide where to put this (ie against the thing, or a separate interface, or the config-description, or...).

How would you suggest we proceed? I'm very happy to look at alternatives, and very happy to help implementing things, but I would like to move it forward reasonably soon...

Cheers
Chris

dnobel commented:
> How would you suggest we proceed? I'm very happy to look at alternatives, and very happy to help implementing things, but I would like to move it forward reasonably soon...

We should continue the discussion based on your proposal. I will talk to the colleagues to find out how to proceed.

dnobel commented:
Hi Chris,

just to keep you up to date. The discussion brought up some general problems with the current configuration mechanism. I think it also fits to your point, that you want a callback just for a single value which has been changed. So actually we are not sure, if the config action you proposed just tries to fix the problem with the current concept, or if it really is a new concept that is needed.

Maybe we should differentiate between thing handler configuration (polling interval, ip address) and internal device configuration (ramp-up-time, motion sensor sensitivity). Internal device configuration should not be stored in Eclipse SmartHome and must be readout dynamically from the thing. In addition you would expect a separate callback like "updateConfiguration(Map<String, Object> changedProperties)", where you only get the changed values. We are thinking about changing the concepts and APIs to have these configuration splitted.

In addition to that we may need a concept for stateless methods (or actions) like a reset on the device. Maybe these methods can be parametrized. But I don´t know if this should be part of the configuration and if we really need all the meta data like options. Maybe we could also model it as a special channel.

One last question to your current requirements:

* Is the main difference for you between a configuration parameter and a config action, that a config action does not have value to display?
* Would it also fulfill your requirement to have method call with typed parameters (string, int, boolean), but no extra meta data like options, min, max etc?

e.g:

```xml
<thing-type id="someType">
        <!-- ... -->
        <action name="refresh">
               <parameter name="a" type="integer" />
               <parameter name="b" type="boolean" />
        </action>
</thing-type>
```

Regards

Dennis

cdjackson commented:
Hi Dennis,

> Maybe we should differentiate between thing handler configuration (polling interval, ip address) and internal device configuration (ramp-up-time, motion sensor sensitivity).
This might be interesting. Device configuraiton could always be sent with as an 'action' (ok, maybe we redefine terminology, but I mean actions as we are discussing here).

However...


> Internal device configuration should not be stored in Eclipse SmartHome and must be readout dynamically from the thing. 

This is not a good idea. I give you an example. In zwave, for battery devices they only wake up every few hours, or maybe even every day. So, you can only read configuration back when they wake up (they may wake up in between this to send status - eg motion - but you can't read configuration). So, it means that you can not know the device configuration 'out of hours'. I think ESH needs to persist this information.
However, this is linked to a previous concept you rejected with respect to adding a "PENDING" flag to configuration items (see https://www.eclipse.org/forums/index.php/t/1066042/). This is the reason I wanted this - it allows us to know that the value provided by ESH is currently being updated when the device wakes up next.
I have this concept in OH1 with HABmin/ZWave and it's very useful for zwave at least.

> In addition you would expect a separate callback like "updateConfiguration(Map changedProperties)", where you only get the changed values. We are thinking about changing the concepts and APIs to have these configuration splitted.

At first thought, this probably sounds ok...

> In addition to that we may need a concept for stateless methods (or actions) like a reset on the device. Maybe these methods can be parametrized. But I don´t know if this should be part of the configuration and if we really need all the meta data like options. Maybe we could also model it as a special channel.

I'm also not sure that we need parameters, but I would consider implementing a reasonably rich interface to avoid changes later - or at least provide an interface that allows further enhancements later.

> Is the main difference for you between a configuration parameter and a config action, that a config action does not have value to display?

I don't think so. I think the single main difference would be that an action is sent immediately. It allows updates without change (as per my previous example where we want to send the same value without changing to a different value). However, you are right that in many/most cases, my definition of an action probably wouldn't have a value to display.


> Would it also fulfill your requirement to have method call with typed parameters (string, int, boolean), but no extra meta data like options, min, max etc?

Yes, this would be ok for me, but I do fear that this might not be sufficient in future (ie if/when someone more troublesome than me comes along :) ). For me, it's ok though.

Cheers
Chris

rodneyrehm commented:
Hey guys,

(I'm the non-ESH-contributing party working with @AlexTugarev on defining a way to configure Things. My perspective is that of the user interaction, not the technical nitty gritty of ESH. I'm not involved in ESH in any way.)

> In zwave, for battery devices […]

same situation with HomeMatic and ZigBee.

> […] a previous concept you rejected with respect to adding a "PENDING" flag to configuration items

As far as I understood the thing status management stuff recently added to ESH, we get a general `CONFIGURATION_PENDING` status on Thing level. That does, however, not identify the specific parameters we're waiting for to be synced. None of our (Qivicon) concepts require this level of detail for now.

> I think ESH needs to persist this information.

Well, at the very least cache it. I'm with Dennis on the front that the device-internal configuration might better be kept separate from the ESH-internal configuration. The latter config is required to establish communication with the device, the former to change the device's behavior - entirely different goals. The binding must understand the semantics of ESH-internal config, because it actually needs to work with that data. For the device-internal configuration that is not necessarily true, as config might simply be passed-through.

cdjackson commented:
> As far as I understood the thing status management stuff recently added to ESH, we get a general CONFIGURATION_PENDING status on Thing level. That does, however, not identify the specific parameters we're waiting for to be synced. None of our (Qivicon) concepts require this level of detail for now.

Yes, that's correct. As per the discussion on this previously, this is better than nothing, but when things take maybe a day to update, it would be nice to have finer resolution - ie to know what parameters haven't yet been updated rather than just knowing that *something* hasn't been updated. That's my opinion, and I implemented this in OH1 for zwave as I was getting a lot of questions like "I updated this in the UI hours ago and it's still not working". Allowing the user to understand what paramters have been updated, and what ones are still waiting was really useful.

>Well, at the very least cache it. I'm with Dennis on the front that the device-internal configuration might better be kept separate from the ESH-internal configuration. 

Agreed - I'm not against separating them. That said, from the user perspective that can be as simple as putting the configuration into different configuration groups which was my plan.

The issue though is that I'd hate for the data to not be persisted as it is now - this would be a problem for zwave users as it might take a day or two to reload everything. Alternatively, I could implement a separate persistence for device configuration as is done in OH1 for zwave, but this seems a bit silly to implement something in a binding like this that (IMHO) probably should be provided by the framework (and it is now).

rodneyrehm commented:
> to know what parameters haven't yet been updated rather than just knowing that something hasn't been updated

I don't disagree, but we felt comfortable enough with a device-specific state to not over-complicate state management right from the start. That said, we'll run into a need for something similar, soon. The goal is to allow the binding to mark the parameters or parameterGroup that needs the user's attention to get the device working again - i.e. the binding knows exactly when the credentials didn't work as opposed to when it couldn't connect to some IP in the first place. But let's not get ahead of ourselves and try taking this configuration thing step by step.

>  That said, from the user perspective that can be as simple as putting the configuration into different configuration groups which was my plan.

I don't think that'll fly in our case. The groups are going to be used to aggregate parameters that need to be configured as a set. I think we should avoid conflicting semantics like that.

> The issue though is that I'd hate for the data to not be persisted as it is now - this would be a problem for zwave users as it might take a day or two to reload everything.

In this case you're talking about a potentially stale cache. I don't think this will be a real problem (at least it hasn't been in the past). I agree, persisting the device-config is necessary. I'd just try to keep that separate from the thing configuration.

cdjackson commented:
> In this case you're talking about a potentially stale cache.

True, but at least it provides the last known state. Given that most of this stuff doesn't update often, it's probably right, and (IMHO) you're better to get told what the state was last time it was accessed than just not to have any indication. For zwave configuration (and probably most other devices) the configuration doesn't randomly change - it will only be changed by ESH, so therefore the stale cache problem is (mostly) not an issue.

No problem from my side on your other comments...

dnobel commented:
Hi Chris,

sorry for coming back so late. But today I had a discussion with Kai about this topic. Let me summarize the outcome:

We think we don´t need a new concept for config actions, because your main requirement is to have callback in the binding, which is called explicitly with the changed parameter(s). As discussed before the current configuration concept has some problems, which we want want to fix. Our idea is to leave the XML as it is. But the way how configuration comes into the binding will be changed:

The framework gets a new REST endpoint `things/{thindUID}/config`, where the UI can send a map of parameters. e.g.

```json
PUT things/hue:bridge:13832237/config
{
"secret": "somevalue",
"ip": "somevalue"
}
```

In addition we want to introduce a new callback in the ThingHandler named `configurationUpdated(Map<String, Object> parameters)`, which will be called by the new REST endpoint. In this method the binding only gets the transmitted values, even if the value has not been changed. Moreover the binding is now responsible for validating configuration values and persisting the parameters. The default implementation of the method could look like this:

```java
public void configurationUpdated(Map<String, Object> changedParameters) throws Exception {
        Configuration configuration = editConfiguration();
        Set<Entry<String, Object>> entrySet = changedParameters.entrySet();
        for (Entry<String, Object> entry : entrySet) {
            configuration.put(entry.getKey(), entry.getValue());
        }
        updateConfiguration(editConfiguration);
        dispose();
        initialize();
}
```

UI´s can decide if they want to render a form for all configuration parameters or just submit each parameter individually. The only exception here is the parameter group you introduced. We want to have a fixed semantic, that all parameters in a group must be submitted together. If there is a strong need from your side to distinguish between "action parameters", which should have a special meaning, we may introduce a new attribute in the XML called something like "atomic" (name tbd), which indicates that this parameter should be rendered and submitted individually and not together with other fields.

For real device actions/functions (device functionality, which does not have a value like a "reset" function), we would like you to ask if we could shift this feature to the future. For your current z-wave binding you could model it with a configuration parameter or a channel. Moreover you could also use a specific "hidden" parameter, which only HABmins knows, for example (if the changedParameters map contains a value "reset", you can trigger the reset).

So what do you think about this proposal? Does it fulfill your requirements?

Regards

Dennis

cdjackson commented:
Hi Dennis,

> We think we don´t need a new concept for config actions, because your main requirement is to have callback in the binding,
> 
That’s fine - I think I also had that as a possible option.


> But the way how configuration comes into the binding will be changed:
> 
Again, that looks good…


>  The only exception here is the parameter group you introduced. We want to have a fixed semantic, that all parameters in a group must be submitted together.
> 
Why do you want to set this limitation?

I think it is unnecessary, and potentially limiting. As I’ve said, for zwave it might take some time to set the config and if we have a lot of parameters to set, then this isn’t ideal. If you really want to have this function, then why not add an attribute to the group to enable this rather than requiring it at all times?


> If there is a strong need from your side to distinguish between "action parameters", which should have a special meaning, we may introduce a new attribute in the XML called something like "atomic" (name tbd), which indicates that this parameter should be rendered and submitted individually and not together with other fields.
> 
I don’t think I need this - unless of course you insist that all parameters in a group are sent together - then this might be useful to undo this constraint :)


> For real device actions/functions (device functionality, which does not have a value like a "reset" function), we would like you to ask if we could shift this feature to the future.
> 
That’s Ok if you prefer to leave it (for now). As you say, so long as we have the ability to send a parameter update via a callback, then I can still do what I need even if it’s not ideal.

What about the idea of adding to the TYPE attribute of the parameter to add such a feature? That way the UI can render this appropriately, but we don’t really need (I think) to change the XML structure or the callback.

dnobel commented:
Hi Chris,

Thanks for the feedback, please find my comments below:

>> The only exception here is the parameter group you introduced. We want to have a fixed semantic, that all parameters in a group must be submitted together.

>Why do you want to set this limitation?

> I think it is unnecessary, and potentially limiting. As I’ve said, for zwave it might take some time to set the config and if we have a lot of parameters to set, then this isn’t ideal. If you really want to have this function, then why not add an attribute to the group to enable this rather than requiring it at all times?

We have use cases, where it only makes sense to set parameters together, e.g: credentials like username and password, or WIFI credentials (SSID + passphrase). And because we already have the parameter group, we thought that if the UI renders the fields together it would be a good idea that the UI also sends the parameters together. 

Before we introduce a new attribute, do you see use cases, where it don´t makes sense that the properties of a group are sent together? If so, we could add an attribute to the group.

>> If there is a strong need from your side to distinguish between "action parameters", which should have a special meaning, we may introduce a new attribute in the XML called something like "atomic" (name tbd), which indicates that this parameter should be rendered and submitted individually and not together with other fields.
>I don’t think I need this - unless of course you insist that all parameters in a group are sent together - then this might be useful to undo this constraint :)

I tend to remove the current constraint, to not add new attribute :)

>>For real device actions/functions (device functionality, which does not have a value like a "reset" function), we would like you to ask if we could shift this feature to the future.
>That’s Ok if you prefer to leave it (for now). As you say, so long as we have the ability to send a parameter update via a callback, then I can still do what I need even if it’s not ideal.

Good, thanks!

> What about the idea of adding to the TYPE attribute of the parameter to add such a feature? That way the UI can render this appropriately, but we don’t really need (I think) to change the XML structure or the callback.

What do you mean exactly? Adding a type to a parameter? There already is a type: string, boolean or decimal. In addition we have the context to refine the meaning of the parameter.

If it is OK for you, I would like to close this PR and I will start the implementation of the discussed concept. Maybe we can come back to this PR, if we later want to introduce real actions. Again sorry, that the discussion took so long and was a bit confusing.

Regards

Dennis

cdjackson commented:
> We have use cases, where it only makes sense to set parameters together, e.g: credentials like username and password, or WIFI credentials (SSID + passphrase). And because we already have the parameter group, we thought that if the UI renders the fields together it would be a good idea that the UI also sends the parameters together.
> 
Sounds fine - I agree for use cases like this make sense.

> Before we introduce a new attribute, do you see use cases, where it don´t makes sense that the properties of a group are sent together? If so, we could add an attribute to the group.
> 
Not that it doesn’t “make sense” - it wouldn’t be ‘destructive’ to send all parameters, but more that it’s unnecessary and could cause problems. Again using zwave as the example, we might have (say) 10 parameters in a group, but they don’t need to be updated together, and if I update all parameters, then it just takes time (potentially hours due to the wakeup interval in zwave!), and it’s not necessary to do this.

> What do you mean exactly? Adding a type to a parameter? There already is a type: string, boolean or decimal. In addition we have the context to refine the meaning of the parameter.
> 
What I meant was (for example) to add a new type - e.g. type ‘action’ (maybe not ‘action', but maybe toggle, or something else) that means that this is rendered as a button or menu, or ? in the UI… It was just a thought…

> If it is OK for you, I would like to close this PR and I will start the implementation of the discussed concept. Maybe we can come back to this PR, if we later want to introduce real actions. Again sorry, that the discussion took so long and was a bit confusing.
> 
Yep - that’s fine - thanks.

dnobel commented:
Hi Chris,

as discussed I will close this PR now. I created a ticket in the issue tracker: https://bugs.eclipse.org/bugs/show_bug.cgi?id=471050 which summarizes the discussion and can be used as follow-up.

Unfortunately I will not have the time to implement the discussed solution in the next weeks. If you want to implement it, feel free to start. Otherwise, if you can wait 2-3 weeks I will implement it.

Regards

Dennis

dnobel commented:
@cdjackson Just want to let you know, there I´m already working on that issue. I hope to come up with a PR soon.
